cpp/src/parquet/chunker_internal.cc

```diff
@@ -0,0 +1,413 @@
+// @return The mask used to compare against the rolling hash
+static uint64_t CalculateMask(int64_t min_chunk_size, int64_t max_chunk_size,
+                              int8_t norm_factor) {
+  if (min_chunk_size < 0) {
+    throw ParquetException("min_chunk_size must be positive");
+  }
+  if (max_chunk_size <= min_chunk_size) {
+    throw ParquetException("max_chunk_size must be greater than min_chunk_size");
+  }
+
+  // calculate the average size of the chunks
+  int64_t avg_chunk_size = (min_chunk_size + max_chunk_size) / 2;
+  // since we are skipping the first `min_chunk_size` bytes for each chunk, we need to
+  // target a smaller chunk size to reach the average size after skipping the first
+  // `min_chunk_size` bytes
+  int64_t target_size = avg_chunk_size - min_chunk_size;
+
+  // assuming that the gear hash has a uniform distribution, we can calculate the mask
+  // by taking the floor(log2(target_size))
+  int mask_bits = std::max(0, ::arrow::bit_util::NumRequiredBits(target_size) - 1);
+
+  // 3 because we are using 8 hash tables to have more gaussian-like distribution,
+  // a user defined `norm_factor` can be used to adjust the mask size, hence the matching
+  // probability, by increasing the norm_factor we increase the probability of matching
+  // the mask, forcing the distribution closer to the average size; norm_factor is 0 by
+  // default
+  if (norm_factor < -3 || norm_factor > 3) {
+    ARROW_LOG(WARNING) << "norm_factor=" << std::to_string(norm_factor)
+                       << " is outside the recommended range [-3, 3]";
+  }
+
+  int mask_adjustement = 3 + norm_factor;
+  int effective_bits = mask_bits - mask_adjustement;
+  if (effective_bits == 0) {
+    return 0;
+  } else if (effective_bits > 64) {
+    throw ParquetException("The number of bits in the mask cannot exceed 64, got " +
+                           std::to_string(effective_bits));
+  } else if (effective_bits < 0) {
+    throw ParquetException(
+        "The difference between min_chunk_size=" + std::to_string(min_chunk_size) +
+        " and max_chunk_size=" + std::to_string(max_chunk_size) +
+        " is too small for the given norm_factor=" + std::to_string(norm_factor) +
+        ", either increase the size range or decrease the norm_factor.");
+  } else {
+    // create the mask by setting the top mask_bits bits
+    return std::numeric_limits<uint64_t>::max() << (64 - effective_bits);
+  }
+}
+
+class ContentDefinedChunker::Impl {
+ public:
+  Impl(const LevelInfo& level_info, int64_t min_chunk_size, int64_t max_chunk_size,
+       int8_t norm_factor)
+      : level_info_(level_info),
+        min_chunk_size_(min_chunk_size),
+        max_chunk_size_(max_chunk_size),
+        rolling_hash_mask_(CalculateMask(min_chunk_size, max_chunk_size, norm_factor)) {}
+
+  uint64_t GetRollingHashMask() const { return rolling_hash_mask_; }
+
+  void Roll(const bool value) {
+    if (chunk_size_++ < min_chunk_size_) {
```

Comment:
Is `chunk_size_++` correct here?
Other `Roll()` updates `chunk_size_` before comparison.

Reply:
Nice catch, it should be pre-incremented, fixed.